---
title: " Cluster Anotation"
output: html_document
date: "2024-05-10"
---

## Overview
The purpose of the following code is to annotate each group obtained through unsupervised clustering using specific databases. Cluster annotation is the process of assigning biological meaning to clusters obtained from unsupervised analysis. When using a reference database like ImmGen, which contains gene expression profiles for various immune cell types, we can identify the cell types present in our data by comparing the expression patterns of our clusters to known profiles in the database.

# Libraries
We start by loading the libraries:
```{r}
library(pheatmap)
#BiocManager::install("celldex")
library(celldex)
library(Seurat)
library(ggplot2)
library(tidyverse)
library(dplyr)
#BiocManager::install("SingleR")
library(SingleR)
#install.packages("remotes")
#remotes::install_github("guokai8/rcellmarker")
library(rcellmarker)
library(ggrepel)
#referencia=celldex::ImmGenData()

```

# ImmGenData

We perform cell annotation of the sorted cells using references from the **ImmGenData** database. ImmGen (Immunological Genome Project) is a comprehensive database that provides high-resolution gene expression profiles for over 250 immune cell types in mice. Using it for cluster annotation helps identify and characterize the immune cell populations in single-cell RNA-seq data, making it easier to understand the biological roles of each cluster.

In summary, ImmGen helps to systematically classify the clusters into meaningful cell types, enhancing the biological interpretation of the data.

The slot we use to anotate is "counts".
```{r, eval=FALSE}

sorted_counts=GetAssayData(datos.integrated, slot="counts") 


```

# SingleR

SingleR is a computational tool used for automated annotation of single-cell RNA-seq data. It assigns cell labels to individual cells based on reference datasets containing well-characterized cell types. SingleR compares the gene expression profiles of your cells to the profiles of known cell types, providing accurate and reproducible annotations.

```{r, eval=FALSE}
pred_sorted= SingleR(test=sorted_counts, ref=referencia, label=referencia$label.main)
table(pred_sorted$labels)
```
These are the number of cells identified by SingleR for each cell group. Some of them, due to their very low count, are negligible: Epithelial cells, mast cells, microglia, stromal cells, as well as T cells.


 | SingleR Immgen        | Counts |
 |-----------------------|--------|
 | B cells               | 726    |
 | B cells, pro          | 256    |
 | Basophils             | 275    |
 | DC                    | 75     |
 | Endothelial cells     | 57     |
 | Eosinophils           | 338    |
 | Epithelial cells      | 1      |
 | ILC                   | 1      |
 | Macrophages           | 727    |
 | Mast cells            | 4      |
 | Microglia             | 1      |
 | Monocytes             | 8202   |
 | Neutrophils           | 15148  |
 | Stem cells            | 768    |
 | Stromal cells         | 9      |
 | T cells               | 4      |

 Let´s create a plot to visualize the distribution of each cell type:

```{r, eval=FALSE}
datos.integrated$SingleR.labels=pred_sorted[match(rownames(datos.integrated@meta.data), rownames(pred_sorted)), "labels"]
View(datos.integrated@meta.data)
DimPlot(datos.integrated, reduction= "umap", group.by = "SingleR.labels", label = T, label.size = 3, repel = TRUE) +
  theme(text = element_text(size = 10))
```
![Cell type Anotation](datos.integrated.sorted/singleRanotacion.png)
# Evaluation of the Anotation

 -**Heatmap**
Now we need to evaluate how well the cells have been annotated.  
To do this, we use a heatmap to visually check the results.


```{r, eval=FALSE}
pred_sorted$scores
plotScoreHeatmap(pred_sorted)
```
![Heatmap of the Anotation´s scores.](datos.integrated.sorted/scoresAnotacionSingleRheatmap.png)
Explanation of the heatmap: Each column represents cells, and the rows represent the labels. The scores indicate how these cells are labeled relative to other labels. The higher the score, the more yellow it appears; the lower the score, the more blue it is.

-**DeltaValues**

We can display and verify it further using **deltaValues**. Deltavalues are a metric used to assess the quality of cell type annotations in single-cell RNA-seq data. In the context of tools like SingleR, delta values provide a quantitative measure of the confidence or reliability of the assigned labels for each cell.
```{r, eval=FALSE}
plotDeltaDistribution(pred_sorted)
```

![](datos.integrated.sorted/DeltaValueAnotacionSingleR.png)
As we can see, in general, the anotation has been quite effective.


**Summary table**
We create tables with the number of cells per cluster according to the cell group.

```{r, eval=FALSE}
tab_sorted= table(Assigned= pred_sorted$labels, Cluster=datos.integrated$seurat_clusters)
tab_sorted=as.data.frame(tab_sorted)
write_xlsx(tab_sorted, "celsXcluster.xlsx")
```

(Note: the table has been saved in a excel format and not is represented here due to the dig dimensions of the table)

We observe that there are clusters containing multiple cell types. This is because it is an unsupervised cluster, which helps us identify new characteristics that may be beneficial for our analysis.

## Other ways to check Anotation.

**1. MouseRNAseqData**
What we do now is compare our results with another dataset, **MouseRNAseqData**. While **ImmGenData** is based on microRNA, **MouseRNAseqData** relies on RNA.

```{r, eval=FALSE}
 referencia2 <- celldex::MouseRNAseqData()
  referencia2$label.main
```


```{r, eval=FALSE}
pred_mouse= SingleR(test=sorted_counts, ref=referencia2, labels=referencia2$label.main)
```

```{r, eval=FALSE}
table(pred_mouse$labels)
```
| MouseRNAseqData       | Counts |
|-----------------------|--------|
| B cells               | 957    |
| Dendritic cells       | 105    |
| Endothelial cells     | 63     |
| Epithelial cells      | 1      |
| Erythrocytes          | 33     |
| Granulocytes          | 15722  |
| Hepatocytes           | 1      |
| Macrophages           | 602    |
| Microglia             | 1      |
| Monocytes             | 9100   |
| NK cells              | 4      |
| T cells               | 3      |


 | SingleR Immgen        | Counts |
 |-----------------------|--------|
 | B cells               | 726    |
 | B cells, pro          | 256    |
 | Basophils             | 275    |
 | DC                    | 75     |
 | Endothelial cells     | 57     |
 | Eosinophils           | 338    |
 | Epithelial cells      | 1      |
 | ILC                   | 1      |
 | Macrophages           | 727    |
 | Mast cells            | 4      |
 | Microglia             | 1      |
 | Monocytes             | 8202   |
 | Neutrophils           | 15148  |
 | Stem cells            | 768    |
 | Stromal cells         | 9      |
 | T cells               | 4      |


There is a certain similarity between both datasets, leading us to believe that **SingleR** is annotating effectively.

Let´s do a graph to check it too:
```{r, eval=FALSE}
datos.integrated$MouseRNA=pred_mouse$labels[match(rownames(datos.integrated@meta.data), rownames(pred_mouse))]
DimPlot(datos.integrated, reduction= "umap", group.by = "MouseRNA", label = T)
```
![MouseRNAseqData Anotation](datos.integrated.sorted/MouseRNAanotacion.png)

As we can see, if we compare both graphs, Immgen and MouseRNA, they both anotate the cell type in a similar way.



**2. Manual Anotation **
We start by searching for the identity markers of the different cell groups to verify that our markers are indeed reflected in the locations defined by **SingleR** for each cell type. This will be determined for the **SHAM** condition.

```{r, eval=FALSE}
sham.sorted <- subset(datos.integrated, subset = condition == "SHAM")

```


```{r,  eval=FALSE}
Idents(datos.integrated)="integrated_snn_res.0.8"

# Check the first values of identities

#head(Idents(datos.integrated))

# We need a summary of the data
table(Idents(datos.integrated))
#Total number of cells per assigned cluster

```

**-FindALLMarkers**

Now we need to get  the markers from the cells. To get the most relevants markers we use the following parameters:

```{r, eval=FALSE}

DefaultAssay(sham.sorted)="RNA"
sham.markers.sorted= FindAllMarkers(sham.sorted, only.pos=TRUE, min.pct = 0.1, 
                                    logfc.threshold = 0.3, 
                            max.cells.per.ident = 2000, slot="counts")
sham.significant_markers <- sham.markers.sorted %>%
  dplyr::filter(p_val_adj < 0.05, avg_log2FC > 0.3) %>%
  group_by(cluster)
```

```{r, eval=FALSE}

write_xlsx(sham.significant_markers, "sham.markers.sorted.xlsx")
```

**-Mapping markers**
Next step is mapping the markers we found in the literate with the SHAM seurat object:
```{r,  eval=FALSE}

#Basophil markers

FeaturePlot(sham.sorted, features=c("Gp9","Hgf"), min.cutoff = "q10", label = TRUE )
FeaturePlot(sham.sorted, features=c( "Cd200r3"), min.cutoff = "q10", label = TRUE )


```
GP9 :9,14
Cd200r3: 19
Hgf: 4, 6


```{r, eval=FALSE}
#Bcells markers
FeaturePlot(sham.sorted, features=c("Cd79b","Cd79a"), min.cutoff = "q10")
FeaturePlot(sham.sorted, features=c("Ly6d"), min.cutoff = "q10", label=TRUE)

```
Ly6d: 13,16
Cd79b:13,16
Cd79a: 13,16

```{r,eval=FALSE}
#macrophage, monocyte markers
FeaturePlot(sham.sorted, features= c("Csf1r", "Klf4"), min.cutoff = "q10")
FeaturePlot(sham.sorted, features= "Cd68")
FeaturePlot(sham.sorted, features=c( "Klf4","Apoe"), min.cutoff = "q10")
FeaturePlot(sham.sorted, features=c("LY62C","Cd68","Cx3cr1"), min.cutoff = "q10")


```
Csf1r. 2, 4, ,6 , 11
Klf4:2 4, ,6 , 11,17,18
Apoe: 2,4, ,6 ,11
Cd68:2, 4, ,6 ,11, 17, 18
Cx3cr1:2,4, 6, 11

```{r, eval=FALSE}
#Dendritic cells markers 
FeaturePlot(sham.sorted, features=c("H2-Ab1", "H2-Aa"), min.cutoff = "q10" )
FeaturePlot(sham.sorted, features=c("Cd209a"), min.cutoff = "q10" )
FeaturePlot(sham.sorted, features=c("H2-Eb1"), min.cutoff = "q10")

```
H2-Ab1:18
H2-Aa:18
Cd209a:18

```{r, eval=FALSE}
#neutrophils markers

FeaturePlot(sham.sorted, features=c("Ly6g", "Cxcr2","S100a8"), min.cutoff = "q10" )
```
Ly6g:7,10,15,20
Ly6g5b: 7,15,20
Ly6g6f, Ly6g6c:9, 14
Cxcr2: 0,1,5,7,10,12

```{r, eval=FALSE}
#Natural killers markers
FeaturePlot(sham.sorted, features=c("Gzma", "Ccl5"), min.cutoff = "q10" )
FeaturePlot(sham.sorted, features=c("Klrd1", "Nkg7"), min.cutoff = "q10" )

```
Klrd1:18
Nkg7:20
Ccl5:11


```{r, eval=FALSE}
#tcells markers
FeaturePlot(sham.sorted, features=c("Cd3e", "Cd4","Cd8a"), min.cutoff = "q10")
FeaturePlot(sham.sorted, features=c( "Cd6", "Cd27","TraC","Trbc2"), min.cutoff = "q10" )

```
CD3eap: 2, ,4 , 11, 17, 18
CD48:2,4,6,11,18
CD47:2,4,6,7,11,17,18,20
CD44:2,4,11,17
CD40:11,18

CD68:2,4,6,11,17,18
CD63:7,8, 10,15,17,20
CD69:13,16
CD27:13,16

```{r, eval=FALSE}
#Endothelial cells markers
FeaturePlot(sham.sorted, features=c("Pecam1", "Eng"), min.cutoff = "q10" )
FeaturePlot(sham.sorted, features=c("Cd34", "Vwf"), min.cutoff = "q10" )
FeaturePlot(sham.sorted, features=("Cdh5"), min.cutoff = "q10")
FeaturePlot(sham.sorted, features=("Cd34"), min.cutoff = "q10")
FeaturePlot(sham.sorted, features=("Kdr"), min.cutoff = "q10")
FeaturePlot(sham.sorted, features=("Flt1"), min.cutoff = "q10")
FeaturePlot(sham.sorted, features=("Pecam1"), min.cutoff = "q10")
```
Cd34:21
Vwf:9,14
Pecam1:6, 9,13,14,16,21
Eng:11,16,18,21
#Endothelial cluster: 21


```{r, eval=FALSE}
#Stem cells markers
FeaturePlot(sham.sorted, features=c("Kit"), min.cutoff = "q10")
FeaturePlot(sham.sorted, features=c("Cd34"), min.cutoff = "q10")
```
Kit:16,20,21


```{r, eval=FALSE}
#megakariocits markers
FeaturePlot(sham.sorted, features=c("Ppbp","Pf4"), min.cutoff = "q10" )
```
Ppbp:9,14
Pf4:9,14
(Note: all the numbers behind every code, represents the markers and in which cluster where found )


We represent the manually selected markers alongside the cell types:

**DOTPLOT :**
```{r, eval=FALSE}
# Lista de genes por tipo de célula
genes_to_plot <- c("Ptprc",
                  "Csf1r", "Apoe", "Cd14", "Cd86", "F13a1","Ly6c2",
                  "Cxcr2", "Ly6g", "Ngp", "S100a8",
                  "Cd209a", "H2-Aa", "H2-Ab1",
                  "Cd79a", "Cd79b", "Ly6d",
                  "Pecam1","Cdh5", "Procr", "Eng", "Kdr", "Vwf","Mcam",
                   "Aqp9", "Cd200r3", "Hgf",
                   "Pf4", "Ppbp"  ,                  
                    "Cd34", "Kit" )                      
# Generar el DotPlot
DotPlot(object = datos.integrated, features = genes_to_plot) +
  RotatedAxis() +  # Rotar las etiquetas de los genes para mejor visualización
  theme(axis.text.x = element_text(size = 7, angle = 45, hjust = 1))

# Ejemplo de cómo definir el DotPlot
DotPlot(object = datos.integrated, features = genes_to_plot, group.by = "cellgeneric", split.by = "condition") +
    scale_y_discrete(labels = c("Monocitos_SHAM" = "Mo_SHAM", "Monocitos_HI" = "Mo_HI", 
                              "Monomix_SHAM" = "Mo-mix_SHAM", "Monomix_HI" = "Mo-mix_HI", 
                              "Neutrophils_SHAM" = "Neu_SHAM", "Neutrophils_HI" = "Neu_HI", 
                              "DC_SHAM" = "DC_SHAM", "DC_HI" = "DC_HI",
                              "B cells_SHAM" = "B_SHAM", "B cells_HI" = "B_HI", 
                              "SCmix_SHAM" = "PLmix_SHAM", "SCmix_HI" = "PLmix_HI", 
                              "Endothelial cells_SHAM" = "EC_SHAM", "Endothelial cells_HI" = "EC_HI")) +

  theme_minimal() +
  theme(axis.text.x = element_text(size= 7, angle = 45, hjust = 1)) +  # Rotar etiquetas para visibilidad
  guides(fill = guide_legend(title = "condition")) # Añadir leyenda para el nivel de expresión



      DotPlot(object = datos.integrated, features = genes_to_plot, group.by = "cellgeneric", split.by = "condition") +
  #scale_fill_manual(values = c("SHAM" = "blue", "HI" = "red"), name = "condition", labels = c("SHAM", "HI")) + 
  scale_y_discrete(limits = c("Monocitos_SHAM", "Monocitos_HI",
                              "Monomix_SHAM", "Monomix_HI",
                            "Neutrophils_SHAM", "Neutrophils_HI",
                              "DC_SHAM", "DC_HI",
                              "B cells_SHAM", "B cells_HI",
                              "SCmix_SHAM", "SCmix_HI",
                             "Endothelial cells_SHAM", "Endothelial cells_HI"),
    
                   labels = c("Monocitos_SHAM" = "Mo", "Monocitos_HI" = "Mo", 
                              "Monomix_SHAM" = "Mo-mix", "Monomix_HI" = "Mo-mix", 
                              "Neutrophils_SHAM" = "Neu", "Neutrophils_HI" = "Neu", 
                              "DC_SHAM" = "DC", "DC_HI" = "DC",
                              "B cells_SHAM" = "B", "B cells_HI" = "B", 
                              "SCmix_SHAM" = "SCmix", "SCmix_HI" = "PLmix", 
                              "Endothelial cells_SHAM" = "EC", "Endothelial cells_HI" = "EC")) +
  theme_minimal() +
  theme(axis.text.x = element_text(size=7, angle = 45, hjust = 1)) +  # Rotar etiquetas para visibilidad
  guides(fill = guide_legend(title = "Condition")) 


```
![DotPlot: Markers represented by Cell type by condition ](tablas excel, recuento, x cluster, x condicion, QC/dotPlotconditionBien.png)


The following code creates a DotPlot figure that visualizes the same information but separates the conditions:

```{r, eval=FALSE}
# Filtrar los datos para la condición SHAM
DotPlot(object = subset(datos.integrated, subset = condition == "SHAM"), 
        features = genes_to_plot, 
        group.by = "cellgeneric") +
  scale_y_discrete(
                   labels = c("Monocitos_SHAM" = "Mo", 
                              "Monomix_SHAM" = "Mo-mix", 
                              "Neutrophils_SHAM" = "Neu", 
                              "DC_SHAM" = "DC", 
                              "B cells_SHAM" = "B", 
                              "SCmix_SHAM" = "SCmix", 
                              "Endothelial cells_SHAM" = "EC")) +
  theme_minimal() +
  theme(axis.text.x = element_text(size=7, angle = 45, hjust = 1)) +
guides(color = guide_legend(title = "MIN(gray)-Max(blue)"),
         size = guide_legend(title = "% Expresion"))  +
  labs(title = "SHAM")  

DotPlot(object = subset(datos.integrated, subset = condition == "HI"), 
        features = genes_to_plot, 
        group.by = "cellgeneric") +
  scale_y_discrete(
                      labels=c ("Monocitos_HI" = "Mo", 
                              "Monomix_HI" = "Mo-mix", 
                              "Neutrophils_HI" = "Neu", 
                              "DC_HI" = "DC", 
                              "B cells_HI" = "B", 
                              "SCmix_HI" = "PLmix", 
                              "Endothelial cells_HI" = "EC") )+
  theme_minimal() +
  theme(axis.text.x = element_text(size=7, angle = 45, hjust = 1)) +
guides(color = guide_legend(title = "MIN(gray)-Max(blue)"),
         size = guide_legend(title = "% Expresion"))  +
  labs(title = "HI")  

    

```

![DotPlot: Markers represented by Cell type by HI condition ](tablas excel, recuento, x cluster, x condicion, QC/HIdotplot.png)
![DotPlot: Markers represented by Cell type by SHAM condition ](tablas excel, recuento, x cluster, x condicion, QC/SHAMdotplot.png)

# Manual labels
We rename the clusters based on the data we obtained from the markers. So now we add a column in the metadata with this new classification.

```{r, eval=FALSE}



datos.integrated <- RenameIdents(datos.integrated, `2` = "Monocitos", `4` = "Monocitos", `6` = "Monocitos", `11` = "Monocitos")

datos.integrated <- RenameIdents(datos.integrated, `3` = "Monomix", `19` = "Monomix", `17` = "Monomix", `21` = "Endothelial cells",`13` = "B cells", `16` = "B cells", `0` = " Neutrophils", `1` = " Neutrophils", `7` = " Neutrophils", `5` = " Neutrophils" , `10` = " Neutrophils", `12` = " Neutrophils", `8` = " Neutrophils", `20` = " Neutrophils", `15` = " Neutrophils", `9` = " SCmix", `14` = " SCmix", `18` = " DC" )
datos.integrated$cellgeneric= Idents(object=datos.integrated)

# Check the changes
table(Idents(datos.integrated))


```
```{r, eval=FALSE}
DimPlot(datos.integrated, reduction="umap", label=T)
```
![New manual label: cellgeneric](datos.integrated.sorted/cellgeneric.png)

*Summary table*
We create a summary table of the number of cells according to cell group( cell generic label) and condition.
```{r, eval=FALSE}
celltable <- table(datos.integrated$cellgeneric[datos.integrated$condition=="SHAM"])
celltable <- rbind(celltable,table(datos.integrated$cellgeneric[datos.integrated$condition=="HI"]))
celltable <- rbind(celltable,abs(celltable[2,]-celltable[1,]))
rownames(celltable) <- c("SHAM","HI","Difference")
celltable <- rbind(celltable, Total = colSums(celltable[1:2,]))
celltable= as.data.frame(celltable)
```
```{r, eval=FALSE}
write.xlsx(celltable, file="celsXcellgenericXcondition.xlsx", rowNames= T)
```

	          |Monocitos	|Monomix	|EC  |	B cells	| Neutrophils	| SCmix	| DC  |
|SHAM       |	3682      |	735	    | 18 |  	489	  |   6569      |	720   |	140 |
|HI	        |3605	      |1684     |	37 |	612	    |   7422      |728    |	151 |
|Difference |	77        |	949     |	19 |	123	    |    853      |	8     |	11  |
|Total      |	7287	    |2419     |	55 |	1101    |	  13991	    |1448	  | 291 |


We create other table that reflects numbers of cells of every sample and condition.

```{r, eval=FALSE}
celltableSample<- table(datos.integrated$cellgeneric[datos.integrated$sample=="sm1HI"])
celltableSample<- rbind(celltableSample,table(datos.integrated$cellgeneric[datos.integrated$sample=="sm2HI"]))
celltableSample<- rbind(celltableSample,table(datos.integrated$cellgeneric[datos.integrated$sample=="sm8SHAM"]))
celltableSample<- rbind(celltableSample,table(datos.integrated$cellgeneric[datos.integrated$sample=="sm9SHAM"]))
rownames(celltableSample) <- c("sm1HI","sm2HI","sm8SHAM", "sm9SHAM")

celltableSample= as.data.frame(celltableSample)
```
```{r, eval=FALSE}
write.xlsx(celltableSample, file="celsXcellgenericXSample.xlsx", rowNames= T)
```
We save these tables as excell format.


At the end, we save the workspace as "newAnotation.RData".





